beets/library.py

```diff
@@ -715,6 +715,71 @@ def move(self, copy=False, link=False, basedir=None, with_album=True):
 
     # Templating.
 
+    def _legalize_partial(self, path, fragment, replacements):
+        """ Perform a partial legalization of the path (ie. a single
+        sanitization and truncation). Outputs unicode if fragment is set,
+        otherwise bytes.
+        """
+
+        # Truncate components and remove forbidden characters.
+        path = util.sanitize_path(path, replacements)
+
+        # Encode for the filesystem.
+        if not fragment:
+            path = bytestring_path(path)
+
+        # Preserve extension.
+        _, extension = os.path.splitext(self.path)
+        if fragment:
+            # Outputting Unicode.
+            extension = extension.decode('utf8', 'ignore')
+        path += extension.lower()
+
+        # Truncate too-long components.
+        maxlen = beets.config['max_filename_length'].get(int)
+        if not maxlen:
+            # When zero, try to determine from filesystem.
+            maxlen = util.max_filename_length(self._db.directory)
+        path = util.truncate_path(path, maxlen)
+
+        return path
+
+    def _legalize_path(self, path, fragment):
+        """ Perform several iterations of _legalize_partial, to generate a
+        stable, optimal output path. This is necessary for cases where
+        truncation produces unclean paths (eg. trailing space).
+        """
+
+        # Create a list of path candidates
+        path_candidates = [b'']
+
+        replacements = self._db.replacements
+
+        # Perform an initial pass
+        path = self._legalize_partial(path, fragment, replacements)
+        while path != path_candidates[-1]:
+            # This will keep sanitizing the path until it's stable, or an
+            # infinite loop appears
+            while path not in path_candidates:
```

Comment:
Nifty: you're explicitly detecting an infinite loop by checking whether we're backtracking to a previously-seen state.

An even simpler heuristic would probably work for our purposes: e.g., if after replacing/truncating/replacing in a single cycle the path is too long again. It's less exciting than detecting backtracking, but perhaps this would avoid weird situations where this would cause many iterations.